Faster CopyFromAscii

benaadams opened the pull request:
Copy `ulong` per loop, less variable comparison, less variable changes, split multiblock into own path

Before
```
        Method |                 Type |          Mean |        Median |           RPS |
-------------- |--------------------- |-------------- |-------------- |-------------- |
 OutputHeaders |            Plaintext |   145.7818 ns |   146.7516 ns |  6,859,565.59 |
 OutputHeaders |               Common |   835.2836 ns |   842.3482 ns |  1,197,198.13 |
 OutputHeaders |              Unknown | 1,207.1645 ns | 1,198.2950 ns |    828,387.52 |
```
After
```
        Method |                 Type |          Mean |        Median |           RPS |
-------------- |--------------------- |-------------- |-------------- |-------------- |
 OutputHeaders |            Plaintext |   141.5104 ns |   141.8268 ns |  7,066,619.65 |
 OutputHeaders |               Common |   750.0226 ns |   748.9297 ns |  1,333,293.11 |
 OutputHeaders |              Unknown | 1,089.1873 ns | 1,089.6819 ns |    918,115.78 |
```

Also added test for the function...

mharthoorn commented:
Shift8Idenitity should probably be Shift8Identity

benaadams commented:
PRing a variant of this to coreclr https://github.com/dotnet/coreclr/pull/8969 as then will effect all serialization (json, mvc etc)

cesarbs commented on test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs:
```diff
@@ -1214,6 +1214,71 @@ public void TestGetAsciiStringEscaped(string input, string expected, int maxChar
         }
 
+        [Fact]
+        public unsafe void TestCopyFromAscii()
```
I'd like to see some boundary testing here, since you're copying 8 bytes at a time and then whatever is left.

cesarbs commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs:
```diff
@@ -21,6 +21,9 @@ public struct MemoryPoolIterator
                                                         0x01ul << 48 ) + 1;
 
+        const int Shift16Shift24 = 256 * 256 * 256 + 256 * 256;
+        const int Shift8Idenitity = 256 + 1;
```
As pointed out my @mharthoorn, `Shift8Identity`.

benaadams commented:
Rebased + sp

Need to add boundary testing

benaadams commented:
Added large boundary test range; shared parsing between single and multi block; added 64/32 bit split and alignment processing.

Probably need to rebench

benaadams commented:
Added benchmarks

Before
```
        Method |                 Type |          Mean |     StdDev |           RPS |
-------------- |--------------------- |-------------- |----------- |-------------- |
 OutputHeaders |               Common |   745.1278 ns |  8.6515 ns |  1,342,051.58 |
 OutputHeaders | ContentLengthNumeric |    79.6235 ns |  0.5345 ns | 12,559,109.61 |
 OutputHeaders |  ContentLengthString |    74.5702 ns |  0.7695 ns | 13,410,186.63 |
 OutputHeaders |            Plaintext |   146.7861 ns |  0.9702 ns |  6,812,635.43 |
 OutputHeaders |              Unknown |   412.5315 ns |  5.5656 ns |  2,424,057.36 |
```

After
```
        Method |                 Type |          Mean |     StdDev |           RPS |
-------------- |--------------------- |-------------- |----------- |-------------- |
 OutputHeaders |               Common |   657.4066 ns |  5.1602 ns |  1,521,128.64 |
 OutputHeaders | ContentLengthNumeric |    77.6110 ns |  0.4539 ns | 12,884,770.92 |
 OutputHeaders |  ContentLengthString |    75.5184 ns |  0.6764 ns | 13,241,814.62 |
 OutputHeaders |            Plaintext |   141.2622 ns |  1.8080 ns |  7,079,035.77 |
 OutputHeaders |              Unknown |   368.7510 ns |  4.6133 ns |  2,711,857.20 |
```

benaadams commented:
Added benchmarks

Before
```
        Method |                 Type |          Mean |        Median |           RPS |
-------------- |--------------------- |-------------- |-------------- |-------------- |
 OutputHeaders |               Common |   835.2836 ns |   842.3482 ns |  1,197,198.13 |
 OutputHeaders | ContentLengthNumeric |    79.6277 ns |    79.6925 ns | 12,558,439.91 |
 OutputHeaders |  ContentLengthString |    73.9253 ns |    74.7080 ns | 13,527,176.51 |
 OutputHeaders |            Plaintext |   145.7818 ns |   146.7516 ns |  6,859,565.59 |
 OutputHeaders |              Unknown | 1,207.1645 ns | 1,198.2950 ns |    828,387.52 |
```
After
```
        Method |                 Type |          Mean |        Median |           RPS |
-------------- |--------------------- |-------------- |-------------- |-------------- |
 OutputHeaders |               Common |   750.0226 ns |   748.9297 ns |  1,333,293.11 |
 OutputHeaders | ContentLengthNumeric |    77.5455 ns |    77.7948 ns | 12,895,660.83 |
 OutputHeaders |  ContentLengthString |    74.9223 ns |    75.2811 ns | 13,347,169.60 |
 OutputHeaders |            Plaintext |   141.5104 ns |   141.8268 ns |  7,066,619.65 |
 OutputHeaders |              Unknown | 1,089.1873 ns | 1,089.6819 ns |    918,115.78 |
```

benaadams commented on src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs:
```diff
@@ -21,9 +21,6 @@ public struct MemoryPoolIterator
                                                         0x02ul << 40 |
                                                         0x01ul << 48 ) + 1;
 
-        const int Shift16Shift24 = 256 * 256 * 256 + 256 * 256;
```
https://github.com/dotnet/coreclr/pull/8969#pullrequestreview-17985076
